wiring/inc/spark_wiring_system.h

```diff
@@ -105,6 +105,164 @@ inline SleepResult::SleepResult(WakeupReason r, system_error_t e, pin_t p)
+        config_.wakeup_sources = nullptr;
+    }
+    ~SystemSleepConfiguration() {
+        // Free memory
+        auto wakeupSource = config_.wakeup_sources;
+        while (!wakeupSource) {
+            auto next = wakeupSource->next;
+            delete wakeupSource;
+            wakeupSource = next;
+        }
+    }
+
+    // Getters
+    hal_core_sleep_config_t* halCoreSleepConfig() const {
+        return &config_;
+    }
+
+    SystemSleepMode sleepMode() const {
+        return static_cast<SystemSleepMode>(config_.mode);
+    }
+
+    bool wakeupSourceFeatured(hal_core_wakeup_source_type_t type) const {
+        auto wakeup = config_.wakeup_sources;
+        while (!wakeup) {
+            if (wakeup->type == type) {
+                return true;
+            }
+            wakeup = wakeup->next;
+        }
+        return false;
+    }
+
+    // Setters
+    SystemSleepConfiguration& mode(SystemSleepMode mode) {
+        config_.mode = static_cast<hal_core_sleep_mode_t>(mode);
+        return *this;
+    }
+
+    SystemSleepConfiguration& gpio(pin_t pin, InterruptMode mode) {
+        clearGpioWakeupSources();
+        return addGpioWakeupSource(pin, mode);
+    }
+
+    SystemSleepConfiguration& gpios(std::initializer_list<hal_core_wakeup_source_gpio_t> gpios) {
+        clearGpioWakeupSources();
+        for (auto gpio : gpios) {
+            addGpioWakeupSource(gpio.pin, gpio.mode);
+        }
+        return *this;
+    }
+
+    SystemSleepConfiguration& duration(system_tick_t ms) {
+        // Iterates current wakeup sources to check if it has configured RTC wakeup
+        auto wakeup = config_.wakeup_sources;
+        bool configured = false;
+        while (!wakeup) {
+            if (wakeup->type == HAL_CORE_WAKEUP_SOURCE_TYPE_RTC) {
+                configured = true;
+                break;
+            }
+            wakeup = wakeup->next;
+        }
+        if (configured) {
+            wakeup->source.rtc.ms = ms;
+        } else {
+            auto hal_core_wakeup_source_t wakeupSource = new(std::nothrow) hal_core_wakeup_source_t();
+            if (!wakeupSource) {
+                return *this;
+            }
+            wakeupSource->size = sizeof(hal_core_wakeup_source_t);
+            wakeupSource->version = HAL_CORE_SLEEP_VERSION;
+            wakeupSource->type = HAL_CORE_WAKEUP_SOURCE_TYPE_RTC;
+            wakeupSource->source.rtc.ms = ms;
+            wakeupSource->next = config_.wakeup_sources;
+            config_.wakeup_sources = wakeupSource;
+        }
+        return *this;
+    }
+
+    SystemSleepConfiguration& duration(std::chrono::milliseconds ms) {
+        return duration(ms.count());
+    }
+
+    SystemSleepConfiguration& duration(std::chrono::seconds s) {
+        std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(s);
+        return duration(ms.count());
+    }
+
+    SystemSleepConfiguration& duration(std::chrono::minutes min) {
```

Comment:
There is no need to do this. It's fine to just have one method that takes minimum resolution (in our case milliseconds) and it'll all work out of the box: https://cppinsights.io/s/7dd33ff3